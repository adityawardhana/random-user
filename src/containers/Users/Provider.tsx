import React, { FC, useEffect, useState } from "react";
import qs from "query-string";
import { getUsers } from "../../services/users";
import useFetch from "../../hooks/useFetch";
import useDebounce from "../../hooks/useDebounce";
import UserContext from "./Context";

const Provider: FC<{}> = ({ children }) => {
  const [params, setParams] = useState({
    inc: "gender,name,login,email,registered",
    results: 10,
    page: 1,
    keyword: "",
    sortBy: "",
    sortOrder: "",
    gender: "",
  });
  const [keyword, setKeyword] = useState("");
  const [searchFocus, setSearchFocus] = useState(false);

  const {
    fetch,
    loading,
    error,
    data: { results: users } = {},
  } = useFetch(
    (newParams) =>
      getUsers(qs.stringify(newParams || params, { skipEmptyString: true })),
    { immediate: false }
  );

  const handleChangeSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
    setKeyword(e.target.value);
  };

  const handleSelectGender = (value: string) => {
    setParams({ ...params, gender: value });
  };

  const handleChangeSort = (sortBy: string, sortOrder: string) => {
    setParams({ ...params, sortBy, sortOrder, page: 1 });
  };

  const handleChangePage = (page: number) => {
    setParams({ ...params, page });
  };

  const handleResetFilter = () => {
    setKeyword("");
    setParams({ ...params, page: 1, keyword: "", gender: "" });
  };

  useDebounce(
    () => {
      searchFocus && setParams({ ...params, keyword, page: 1 });
    },
    keyword,
    1000
  );

  useEffect(() => {
    fetch({ ...params });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [params]);

  return (
    <UserContext.Provider
      value={{
        params,
        users,
        loading,
        error,
        keyword,
        setSearchFocus,
        handleChangePage,
        handleChangeSearch,
        handleChangeSort,
        handleSelectGender,
        handleResetFilter,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export default Provider;
