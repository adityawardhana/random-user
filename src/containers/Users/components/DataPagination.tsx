import { Box } from "@adityawardhana/ts-react-ui"
import Pagination from "../../../ui/Pagination";
import useUsers from "../useUser";

const DataPagination = () => {
  const { params, handleChangePage } = useUsers();
  return (
    <Box my={20}>
      <Pagination page={params.page} onChange={handleChangePage} />
    </Box>
  );
};

export default DataPagination;
