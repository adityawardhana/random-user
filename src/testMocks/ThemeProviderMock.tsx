import React from "react";
import theme, { GlobalStyles } from "@adityawardhana/ts-react-ui/lib/theme";
import { ThemeProvider } from "styled-components";

export const setupTheme = (children: React.ReactNode) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};
