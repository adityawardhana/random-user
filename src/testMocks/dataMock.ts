export const resultAPIMock = {
  login: {
    username: "adityawrdhn",
  },
  name: {
    first: "adit",
    last: "w",
  },
  email: "adityawrdhn@gmail.com",
  gender: "male",
  registered: { date: "2003-08-26T21:56:32.575Z" },
};

export const usersMock = {
  params: { sortBy: "email", sortOrder: "ascend", page: 1, gender: "" },
  loading: false,
  users: [...Array(3)].map(() => ({ ...resultAPIMock })),
  keyword: "xxx",
  handleChangeSort: jest.fn(),
  handleChangePage: jest.fn(),
  handleChangeSearch: jest.fn(),
  setSearchFocus: jest.fn(),
  handleSelectGender: jest.fn(),
  handleResetFilter: jest.fn(),
};
