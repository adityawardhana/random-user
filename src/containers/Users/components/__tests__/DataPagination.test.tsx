import { render } from "@testing-library/react";
import { usersMock } from "../../../../testMocks/dataMock";
import { setupTheme } from "../../../../testMocks/ThemeProviderMock";
import Provider from "../../Provider";
import useUsersMock from "../../useUser";
import DataPagination from "../DataPagination";

jest.mock("../../useUser", () => ({
  __esModule: true,
  default: jest.fn(),
}));

describe("Data Pagination unit tests", () => {
  beforeEach(() => {
    (useUsersMock as any).mockReturnValue({ ...usersMock });
  });
  test("should render correctly", () => {
    const { container } = render(
      setupTheme(
        <Provider>
          <DataPagination />
        </Provider>
      )
    );
    expect(container).toMatchSnapshot();
  });
});
