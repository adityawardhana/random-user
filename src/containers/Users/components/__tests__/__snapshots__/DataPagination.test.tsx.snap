// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Pagination unit tests should render correctly 1`] = `
.c0 {
  margin-top: 20px;
  margin-bottom: 20px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c6 {
  font-size: 16px;
  font-weight: 400;
  line-height: 20px;
  margin-right: 8px;
  margin-left: 8px;
  color: #243E47;
  white-space: normal;
}

.c5 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  pointer-events: none;
}

.c5[type="button"] .c4 {
  cursor: pointer;
}

.c5 svg {
  width: 16px;
  height: 16px;
  fill: #A6A8AF;
}

.c7 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  pointer-events: none;
}

.c7[type="button"] .c4 {
  cursor: pointer;
}

.c7 svg {
  width: 16px;
  height: 16px;
  fill: #243E47;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  overflow: hidden;
  border: 1px solid transparent;
  border-radius: 50%;
  border-color: transparent;
  background-color: transparent;
  color: #152E3B;
  padding: 8px;
  height: 28px;
  width: 28px;
  font-family: inherit;
  font-size: 10px;
  font-weight: 500;
  text-align: center;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  box-sizing: border-box;
  -webkit-transition: -webkit-transform 0.1s ease 0s,background 0.3s ease !important;
  -webkit-transition: transform 0.1s ease 0s,background 0.3s ease !important;
  transition: transform 0.1s ease 0s,background 0.3s ease !important;
}

.c2::before {
  content: "";
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c2:hover::before {
  background: rgba(255,255,255,0.1);
}

.c2:focus {
  outline: none;
}

.c2:focus-visible::before {
  background: rgba(50,50,50,0.05);
  -webkit-transition: background 0.1s ease-in-out;
  transition: background 0.1s ease-in-out;
}

.c2:active {
  outline: none;
  -webkit-transform: scale(0.96) !important;
  -ms-transform: scale(0.96) !important;
  transform: scale(0.96) !important;
}

.c2:active::before {
  background: rgba(50,50,50,0.05);
  -webkit-transition: background 0.3s ease;
  transition: background 0.3s ease;
}

.c2:disabled {
  color: #CDD0D7;
  cursor: default;
}

.c2:link,
.c2:visited,
.c2:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2[type="button"],
.c2[type="reset"],
.c2[type="submit"] {
  -webkit-appearance: none;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="c2"
        color="secondary"
        shape="circle"
      >
        <div
          class="c3"
        >
          <s
            style="position: absolute; border-radius: 50%; opacity: 0; width: 35px; height: 35px; transform: translate(-50%, -50%); pointer-events: none; background-color: rgba(255, 255, 255, 0.4);"
          />
        </div>
        <i
          class="c4 c5"
          color="black400"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-chevron-left"
            fill="currentColor"
            height="16"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M9.78 12.78a.75.75 0 01-1.06 0L4.47 8.53a.75.75 0 010-1.06l4.25-4.25a.75.75 0 011.06 1.06L6.06 8l3.72 3.72a.75.75 0 010 1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </i>
      </button>
      <p
        class="c6"
        color="black800"
        type="B16"
      >
        1
      </p>
      <button
        class="c2"
        color="secondary"
        shape="circle"
      >
        <div
          class="c3"
        >
          <s
            style="position: absolute; border-radius: 50%; opacity: 0; width: 35px; height: 35px; transform: translate(-50%, -50%); pointer-events: none; background-color: rgba(255, 255, 255, 0.4);"
          />
        </div>
        <i
          class="c4 c7"
          color="black800"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-chevron-right"
            fill="currentColor"
            height="16"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M6.22 3.22a.75.75 0 011.06 0l4.25 4.25a.75.75 0 010 1.06l-4.25 4.25a.75.75 0 01-1.06-1.06L9.94 8 6.22 4.28a.75.75 0 010-1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </i>
      </button>
    </div>
  </div>
</div>
`;
